crates/sui-core/src/checkpoints2/mod.rs

```diff
@@ -0,0 +1,365 @@
+use crate::checkpoints2::casual_order::CasualOrder;
+use crate::checkpoints2::checkpoint_output::CheckpointOutput;
+pub use crate::checkpoints2::checkpoint_output::LogCheckpointOutput;
+use futures::future::{select, Either};
+use futures::FutureExt;
+use std::collections::HashSet;
+use std::path::Path;
+use std::sync::Arc;
+use sui_metrics::spawn_monitored_task;
+use sui_types::base_types::TransactionDigest;
+use sui_types::error::SuiResult;
+use sui_types::gas::GasCostSummary;
+use sui_types::messages::TransactionEffects;
+use sui_types::messages_checkpoint::{
+    CheckpointContents, CheckpointSequenceNumber, CheckpointSummary,
+};
+use tokio::sync::{oneshot, Notify};
+use tracing::debug;
+use typed_store::rocks::{DBBatch, DBMap};
+use typed_store::traits::TypedStoreDebug;
+use typed_store::Map;
+use typed_store_derive::DBMapUtils;
+
+type CheckpointCommitHeight = u64;
+
+#[derive(DBMapUtils)]
+struct CheckpointStoreTables {
+    /// This table has information for the checkpoints for which we constructed all the data
+    /// from consensus, but not yet constructed actual checkpoint.
+    ///
+    /// Key in this table is the narwhal commit height and not a checkpoint sequence number.
+    ///
+    /// Non-empty list of transactions here might result in empty list when we are forming checkpoint.
+    /// Because we don't want to create checkpoints with empty content(see CheckpointBuilder::write_checkpoint),
+    /// the sequence number of checkpoint does not match height here.
+    pending_checkpoints: DBMap<CheckpointCommitHeight, Vec<TransactionDigest>>,
+
+    /// Maps sequence number to checkpoint contents
+    checkpoint_content: DBMap<CheckpointSequenceNumber, CheckpointContents>,
+
+    /// Maps sequence number to checkpoint summary
+    checkpoint_summary: DBMap<CheckpointSequenceNumber, CheckpointSummary>,
+
+    /// Lists all transaction digests included in checkpoints
+    /// This can be cleaned up on epoch boundary
+    digest_to_checkpoint: DBMap<TransactionDigest, CheckpointSequenceNumber>,
+}
+
+pub struct CheckpointBuilder {
+    tables: Arc<CheckpointStoreTables>,
+    notify: Arc<Notify>,
+    effects_store: Box<dyn EffectsNotifyRead>,
+    output: Box<dyn CheckpointOutput>,
+    exit: oneshot::Receiver<()>,
+}
+
+impl CheckpointBuilder {
+    fn new(
+        tables: Arc<CheckpointStoreTables>,
+        notify: Arc<Notify>,
+        effects_store: Box<dyn EffectsNotifyRead>,
+        output: Box<dyn CheckpointOutput>,
+        exit: oneshot::Receiver<()>,
+    ) -> Self {
+        Self {
+            tables,
+            notify,
+            effects_store,
+            output,
+            exit,
+        }
+    }
+
+    async fn run(mut self) {
+        loop {
+            for (height, roots) in self.tables.pending_checkpoints.iter() {
+                self.make_checkpoint(height, roots)
+                    .await
+                    .expect("Unrecoverable error while making checkpoint");
```

Comment:
Why is this unrecoverable?

Reply:
Error here basically means storage error. I don't think we have any strategy for that so far, so panic seem reasonable

Reply:
It doesn't hurt to come back and retry, right?

Reply:
I am not sure about "it does not hurt". I don't really see conditions where retry could help, and on the other hand panic here allows to surface signal faster. I am also not sure if we have places in code where we retry local errors. On the other hand, we do have places where panic on local storage errors (handling consensus messages is one example)

Reply:
btw the reason this bothers me is that we don't have any panic in the code base today out of storage errors, this would be the first. Also there is no function-level guarantee that make_checkpoint can only return storage error.

Reply:
But we do have places like that, for example ConsensusHandler. Even more so, DBMap itself (in the `Iter::next()`) panics on storage error

Reply:
> On the other hand, we do have places where panic on local storage errors (handling consensus messages is one example)

hmm that's new to me. Well, at least previously we don't. Is the assumption that storage errors are not transient? I am always under the assumption that storage errors could happen transiently.

Reply:
I don't think this is the case. In my view storage error is almost certainly a non-recoverable situation - either faulty disk or out of disk space. I am happy to change my opinion on that if we see storage errors that are recoverable in private testnet

Reply:
I believe you. But as you can tell based on my questions, most of the code has been written in a way that assumes storage errors are somewhat recoverable. More importantly, here my primary worry is one day someone would add a non storage error somewhere that could propagate all the way up.
If you handle the error here by a sleep + continue, it should work?

Reply:
Ok, lets do sleep + continue